e2e-tests/conftest.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2020-2021 by Vytautas Liuolia.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pathlib
+import subprocess
+import time
+
+import pytest
+import requests
+
+import falcon.testing
+
+
+HERE = pathlib.Path(__file__).resolve().parent
+INDEX = '/static/index.html'
+
+
+@pytest.fixture(scope='session')
+def base_url():
+    port = falcon.testing.get_unused_port()
+    base_url = f'http://127.0.0.1:{port}'
+
+    uvicorn = subprocess.Popen(
+        ('uvicorn', '--port', str(port), 'server:app'),
+        cwd=HERE,
+    )
+
+    # NOTE(vytas): give Uvicorn some time to start.
+    for attempt in range(3):
+        try:
+            resp = requests.get(f'{base_url}/ping', timeout=1)
+            resp.raise_for_status()
+            break
+        except requests.exceptions.RequestException:
+            pass
+        time.sleep(attempt + 0.5)
+    else:
+        pytest.fail('Could not start Uvicorn')
+
+    yield base_url
+
+    uvicorn.terminate()
+
+    # NOTE(vytas): give Unicorn a rather generous time period to stop since it
+    #   is waiting for the browser to close open connections such as WS & SSE.
+    uvicorn.communicate(timeout=30)
+
+
+@pytest.fixture()
+def browser(sb, base_url):
```

Comment:
I'm guessing that `sb` is a selenium base fixture.
It may be useful to add a note (and a link to the documentations?) to improve future readability

Reply:
Ah, right, that's a very good comment :+1: 
In fact, their examples are centered around subclassing some sort of `unittest.TestCase` instance, it took me some googling to find out that a pytest fixture is exposed as `sb`. I'll add a note.